File: pe-math/src/main/java/org/saar/maths/objects/Planef.java
package org.saar.maths.objects;

import org.joml.Intersectionf;
import org.joml.Vector3fc;

public class Planef {

    public final float a;
    public final float b;
    public final float c;
    public final float d;

    public Planef(Vector3fc point, Vector3fc normal) {
        this.a = normal.x();
        this.b = normal.y();
        this.c = normal.z();
        this.d = -point.dot(normal);
    }

    public float distance(Vector3fc point) {
        return distance(point.x(), point.y(), point.z());
    }

    public float distance(float px, float py, float pz) {
        return Intersectionf.distancePointPlane(
                px, py, pz, this.a, this.b, this.c, this.d);
    }
}

File: pe-math/src/main/kotlin/org/saar/maths/objects/Planef.kt
package org.saar.maths.objects

import org.joml.*

class Planef(private val normal: Vector3fc, val distance: Float) {

    constructor(normal: Vector3fc, point: Vector3fc) : this(normal, point.dot(normal))

    val a: Float get() = this.normal.x()
    val b: Float get() = this.normal.y()
    val c: Float get() = this.normal.z()

    fun distance(point: Vector3fc): Float {
        return distance(point.x(), point.y(), point.z())
    }

    fun distance(px: Float, py: Float, pz: Float): Float {
        val (a, b, c) = this.normal
        return Intersectionf.distancePointPlane(
            px, py, pz, a, b, c, this.distance
        )
    }
}
